src/hotspot/share/runtime/vmThread.cpp

```diff
@@ -381,119 +287,198 @@ class HandshakeALotClosure : public HandshakeClosure {
   }
 };
 
-VM_Operation* VMThread::no_op_safepoint() {
-  // Check for handshakes first since we may need to return a VMop.
-  if (HandshakeALot) {
-    HandshakeALotClosure hal_cl;
-    Handshake::execute(&hal_cl);
+bool VMThread::handshake_alot() {
+  assert(_cur_vm_operation == NULL, "Already have an op");
+  assert(_next_vm_operation == NULL, "Already have an op");
+  if (!HandshakeALot) {
+    return false;
   }
+  static jlong last_halot_ms = 0;
+  jlong now_ms = nanos_to_millis(os::javaTimeNanos());
+  // If only HandshakeALot is set, but GuaranteedSafepointInterval is 0,
+  // we emit a handshake if it's been more than a second since last.
+  jlong interval = GuaranteedSafepointInterval != 0 ? GuaranteedSafepointInterval : 1000;
+  jlong deadline_ms = interval + last_halot_ms;
+  if (now_ms > deadline_ms) {
+    last_halot_ms = now_ms;
+    return true;
+  }
+  return false;
+}
+
+void VMThread::setup_periodic_safepoint_if_needed() {
+  assert(_cur_vm_operation  == NULL, "Already have an op");
+  assert(_next_vm_operation == NULL, "Already have an op");
   // Check for a cleanup before SafepointALot to keep stats correct.
   long interval_ms = SafepointTracing::time_since_last_safepoint_ms();
   bool max_time_exceeded = GuaranteedSafepointInterval != 0 &&
                            (interval_ms >= GuaranteedSafepointInterval);
-  if (max_time_exceeded && SafepointSynchronize::is_cleanup_needed()) {
-    return &cleanup_op;
+  if (!max_time_exceeded) {
```

Comment:
Hmmm.... The old SafepointALot was intended to safepoint as frequently
as possible to stress the system. Now we do very little at safepoints so
maybe it is time for SafepointALot to evolve. Can you make it so that a
SafepointALot happens some fraction of GuaranteedSafepointInterval, e.g.,
(GuaranteedSafepointInterval / 4) so four times as often?

Reply:
All test using SafepointALot the already set the GuaranteedSafepointInterval to a low value in range of ~1-300ms.
(except for vm boolean flag test which uses SafepointALot  to test a boolean flag)
For example jni/FastGetField sets GuaranteedSafepointInterval to 1.

The only case it would really differ is when adhoc adding SafepointALot without GuaranteedSafepointInterval.

Reply:
If GuaranteedSafepointInterval is set to a lower value than the default on the command line, then I'm okay if SafepointALot does not do anything extra. However, if GuaranteedSafepointInterval is either the default value or is set to a higher value, then I would like SafepointALot to cause a safepoint more frequently than the GuaranteedSafepointInterval. Every GuaranteedSafepointInterval/4 would be a fine definition of "a lot".

Reply:
Mulling on this more... is it too radical to consider that we no longer need SafepointALot?

Reply:
I would like SafepointALot(and HandshakeALot) to be executed in a separate thread and that randomly request a safepoint (preferably with some validation inside the operation).
Since VM thread now handles this it can not do this request while busy.
Also having the VM thread 'more' sporadic waking up will be confusing for the VM thread loop.

So I agree with you that we need a better SafepointALot, but I think it wrong to use the VM thread to drive it.
I suggest we create an enhancement for it.